source/src/microapp/cs_MicroappProtocol.cpp

```diff
@@ -802,19 +321,630 @@ void MicroappProtocol::handleEvent(event_t & event) {
+			}
 			break;
+		}
+		case CS_MICROAPP_COMMAND_PIN_INPUT_PULLUP:
+		case CS_MICROAPP_COMMAND_PIN_READ: {
+			// TODO; (note that we do not handle event handler registration here but in SetMode above
+			break;
+		}
+		default: LOGw("Unknown pin mode / opcode: %i", pin_cmd->opcode2); return ERR_UNKNOWN_OP_CODE;
 	}
+	return ERR_SUCCESS;
+}
+
+int MicroappProtocol::handleMicroappServiceDataCommand(uint8_t* payload, uint16_t length) {
+
+	_logArray(SERIAL_DEBUG, true, payload, length);
+
+	uint16_t commandDataSize = length - 3;  // byte 0 is command, byte 1 is type, byte 2 is option.
+	TYPIFY(CMD_MICROAPP_ADVERTISE) eventData;
+	if (commandDataSize < sizeof(eventData.appUuid)) {
+		LOGi("Payload too small");
+		return ERR_WRONG_PAYLOAD_LENGTH;
+	}
+	eventData.version   = 0;  // TODO: define somewhere.
+	eventData.type      = 0;  // TODO: define somewhere.
+	eventData.appUuid   = (payload[3] << 8) + payload[4];
+	eventData.data.len  = commandDataSize - sizeof(eventData.appUuid);
+	eventData.data.data = &(payload[5]);
+	event_t event(CS_TYPE::CMD_MICROAPP_ADVERTISE, &eventData, sizeof(eventData));
+	event.dispatch();
+	return ERR_SUCCESS;
+}
+
+int MicroappProtocol::handleMicroappTwiCommand(microapp_twi_cmd_t* twi_cmd) {
+	CommandMicroappTwiOpcode opcode = (CommandMicroappTwiOpcode)twi_cmd->opcode;
+	switch (opcode) {
+		case CS_MICROAPP_COMMAND_TWI_INIT: {
+			LOGi("Init i2c");
+			TYPIFY(EVT_TWI_INIT) twi;
+			// no need to write twi.config (is not under control of microapp)
+			event_t event(CS_TYPE::EVT_TWI_INIT, &twi, sizeof(twi));
+			EventDispatcher::getInstance().dispatch(event);
+			break;
+		}
+		case CS_MICROAPP_COMMAND_TWI_WRITE: {
+			LOGd("Write over i2c to address: 0x%02x", twi_cmd->address);
+			uint8_t bufSize = twi_cmd->length;
+			TYPIFY(EVT_TWI_WRITE) twi;
+			twi.address = twi_cmd->address;
+			twi.buf     = twi_cmd->buf;
+			twi.length  = bufSize;
+			twi.stop    = twi_cmd->stop;
+			event_t event(CS_TYPE::EVT_TWI_WRITE, &twi, sizeof(twi));
+			EventDispatcher::getInstance().dispatch(event);
+			break;
+		}
+		case CS_MICROAPP_COMMAND_TWI_READ: {
+			LOGd("Read from i2c address: 0x%02x", twi_cmd->address);
+			uint8_t bufSize = twi_cmd->length;
+			TYPIFY(EVT_TWI_READ) twi;
+			twi.address = twi_cmd->address;
+			twi.buf     = twi_cmd->buf;
+			twi.length  = bufSize;
+			twi.stop    = twi_cmd->stop;
+			event_t event(CS_TYPE::EVT_TWI_READ, &twi, sizeof(twi));
+			EventDispatcher::getInstance().dispatch(event);
+			twi_cmd->ack    = event.result.returnCode;
+			twi_cmd->length = twi.length;
+			// LOGi("Return value of length: %i", twi_cmd->length);
+			// for (int i = 0; i < (int)twi_cmd->length; ++i) {
+			//	LOGi("Read: 0x%02x", twi.buf[i]);
+			//}
+			break;
+		}
+		default: LOGw("Unknown i2c opcode: %i", twi_cmd->opcode); return ERR_UNKNOWN_OP_CODE;
+	}
+	return ERR_SUCCESS;
+}
+
+int MicroappProtocol::handleMicroappBleCommand(microapp_ble_cmd_t* ble_cmd) {
```

Comment:
If you're returning result codes, it should be of type `cs_ret_code_t`
```suggestion
cs_ret_code_t MicroappProtocol::handleMicroappBleCommand(microapp_ble_cmd_t* bleCmd) {
```